File: crackup-php/classes/Crackup.php
<?php
/**
 * Crackup is a static class providing a namespace for Crackup's core
 * functionality.
 * 
 * @static
 * @author Ryan Grove <ryan@wonko.com>
 * @package Crackup
 */
class Crackup {
  // -- Public Static Variables ------------------------------------------------
  public static $driver;
  public static $local       = array();
  public static $localFiles  = array();
  public static $passphrase  = '';
  public static $remote      = '';
  public static $remoteFiles = array();
  
  // -- Public Static Methods --------------------------------------------------
  
  /**
   * Compresses the specified local input file using bzip2 compression and
   * writes the compressed data to the specified output file.
   * 
   * @param string $infile name of the input file
   * @param string $outfile name of the output file
   */
  public static function compressFile($infile, $outfile) {
    $fp = fopen($infile, 'rb');
    $bz = bzopen($outfile, 'w');
    
    while (!feof($fp)) {
      bzwrite($bz, fread($fp, 1048576));
    }
    
    bzclose($bz);
    fclose($fp);
  }
  
  /**
   * Prints the specified debugging message if verbose mode is enabled.
   * 
   * @param string $message message to print
   */
  public static function debug($message) {
    if (defined('VERBOSE')) {
      echo "--> ".$message."\n";
    }
  }
  
  /**
   * Decompresses the specified local input file using bzip2 and writes the
   * decompressed data to the specified output file.
   * 
   * @param string $infile name of the input file
   * @param string $outfile name of the output file
   */  
  public static function decompressFile($infile, $outfile) {
    $bz = bzopen($infile, 'r');
    $fp = fopen($outfile, 'wb');
    
    while(!feof($bz)) {
      fwrite($fp, bzread($bz, 1048576));
    }
    
    fclose($fp);
    bzclose($bz);
  }
  
  /**
   * Decrypts the specified local file using GPG.
   * 
   * @param string $infile filename of the file to decrypt
   * @param string $outfile filename of the file to decrypt to
   */
  public static function decryptFile($infile, $outfile) {
    $gpgCommand = strtr(GPG_DECRYPT, array(
      ':input_file'  => escapeshellarg($infile),
      ':output_file' => escapeshellarg($outfile),
      ':passphrase'  => escapeshellarg(self::$passphrase)
    ));
    
    @shell_exec($gpgCommand);
  }
  
  /**
   * Encrypts the specified local file using GPG.
   *
   * @param string $infile filename of the file to encrypt
   * @param string $outfile filename of the file to encrypt to
   */
  public static function encryptFile($infile, $outfile) {
    $gpgCommand = strtr(GPG_ENCRYPT, array(
      ':input_file'  => escapeshellarg($infile),
      ':output_file' => escapeshellarg($outfile),
      ':passphrase'  => escapeshellarg(self::$passphrase)
    ));
    
    @shell_exec($gpgCommand);
  }
  
  /**
   * Prints the specified error message to stderr and exists with an error code.
   *
   * @param string $message message to print
   */
  public static function error($message) {
    fwrite(STDERR, "Error: ".$message."\n");
    exit(1);
  }
  
  /**
   * Searches the remote file index for a file whose local filename matches the
   * name specified and returns it if found, or <em>false</em> otherwise.
   *
   * @param string $filename local filename to search for
   * @return mixed
   */
  public static function findRemoteFile($filename) {
    $filename = rtrim($filename, '/');
    
    foreach(self::$remoteFiles as $file) {
      if ($file->getName() == $filename) {
        return $file;
      }
      
      if ($file instanceof CrackupDirectory) {
        if ($child = $file->find($filename)) {
          return $child;
        }
      }
    }
    
    return false;
  }
  
  /**
   * Gets an array of CrackupFileSystemObject objects representing the files and
   * directories on the local system in the locations specified by the array of
   * file patterns in Crackup::$local.
   *
   * @return array
   * @see getRemoteFiles()
   */
  public static function getLocalFiles() {
    $localFiles = array();
    
    foreach(self::$local as $pattern) {
      if (false === ($filenames = glob($pattern, GLOB_NOSORT))) {
        continue;
      }
      
      foreach($filenames as $filename) {
        $filename = rtrim($filename, '/');
        
        if (isset($localFiles[$filename])) {
          continue;
        }
        
        if (is_dir($filename)) {
          self::debug($filename);
          $localFiles[$filename] = new CrackupDirectory($filename);
        }
        elseif (is_file($filename)) {
          self::debug($filename);
          $localFiles[$filename] = new CrackupFile($filename);
        }
      }
    }
    
    return $localFiles;
  }
  
  /**
   * Returns an array of CrackupFileSystemObject objects present at the remote
   * location.
   *
   * @return array
   * @see getLocalFiles()
   */
  public static function getRemoteFiles() {
    $tmpFileList = self::getTempFile();
    
    try {
      self::$driver->get(self::$remote.'/.crackup_index', $tmpFileList);
    }
    catch (Exception $e) {
      @unlink($tmpFileList);
      return array();
    }
      
    $oldFile     = $tmpFileList;
    $tmpFileList = self::getTempFile();
    
    if (defined('CRACKUP_NOGPG')) {
      self::decompressFile($oldFile, $tmpFileList);
    }
    else {        
      self::decryptFile($oldFile, $tmpFileList);        
    }
    
    $fileList = @unserialize(@file_get_contents($tmpFileList));
    
    @unlink($oldFile);
    @unlink($tmpFileList);
    
    if (is_array($fileList)) {
      return $fileList;
    }
  
    return array();
  }
  
  /**
   * Gets an array of CrackupFileSystemObject objects representing files and
   * directories that exist on the destination but no longer exist on the local
   * machine.
   *
   * @param array $localFiles array of local CrackupFileSystemObject objects
   * @param array $remoteFiles array of remote CrackupFileSystemObject objects
   * @return array removed CrackupFileSystemObject objects
   * @see getUpdatedFiles
   */
  public static function getRemovedFiles($localFiles, $remoteFiles) {
    $removed = array();
    
    foreach($remoteFiles as $name => $remoteFile) {
      // Remote remote files that no longer exist.
      if (!isset($localFiles[$name])) {
        $removed[] = $remoteFile;
        continue;
      }
      
      $localFile = $localFiles[$name];
      
      if (($remoteFile instanceof CrackupDirectory) && 
          ($localFile instanceof CrackupDirectory)) {
            
        $removed = array_merge($removed, self::getRemovedFiles(
            $localFile->getChildren(), $remoteFile->getChildren()));
      }
    }
    
    return $removed;
  }
  
  /**
   * Generates a unique temporary filename.
   * 
   * @return string
   */
  public static function getTempFile() {
    return TEMP_DIR.'/'.uniqid('.crackup_temp');
  }
  
  /**
   * Gets an array of CrackupFileSystemObject objects representing files and
   * directories that are new or have been modified on the local machine and
   * need to be updated at the destination location.
   *
   * @param array $localFiles array of local CrackupFileSystemObject objects
   * @param array $remoteFiles array of remote CrackupFileSystemObject objects
   * @return array updated CrackupFileSystemObject objects
   * @see getRemovedFiles()
   */
  public static function getUpdatedFiles($localFiles, $remoteFiles) {
    $updated = array();
    
    foreach($localFiles as $name => $localFile) {
      if (!isset($remoteFiles[$name])) {
        $updated[] = $localFile;
        continue;
      }
      
      $remoteFile = $remoteFiles[$name];
      
      if (($localFile instanceof CrackupDirectory) &&
          ($remoteFile instanceof CrackupDirectory)) {

        $updated = array_merge($updated, self::getUpdatedFiles(
            $localFile->getChildren(), $remoteFile->getChildren()));
      }
      elseif (($localFile instanceof CrackupFile) &&
          ($remoteFile instanceof CrackupFile)) {

        if ($localFile->getFileHash() != $remoteFile->getFileHash()) {
          $updated[] = $localFile;
        }
      }
    }

    return $updated;
  }
  
  /**
   * Displays a message and prompts for user input.
   *
   * @param string $message message to display
   * @return string user input
   */
  public static function prompt($message) {
    echo $message.': ';
    return trim(fgets(STDIN));
  }
  
  /**
   * Deletes the specified files/directories from the remote location.
   *
   * @param array $files array of CrackupFileSystemObject objects to delete
   * @see update()
   */
  public static function remove($files) {
    foreach($files as $file) {
      if ($file instanceof CrackupDirectory) {
        self::remove($file);
      }
      elseif ($file instanceof CrackupFile) {
        self::debug($file->getName());
        
        try {
          self::$driver->delete(self::$remote.'/crackup_'.$file->getNameHash());
        }
        catch (Exception $e) {
          self::error('Unable to remove "'.$file->getName().'" from remote '.
              'location');
        }
      }
    }
  }
  
  /**
   * Uploads the specified files/directories to the remote location.
   *
   * @param array $files array of CrackupFileSystemObject objects to update
   * @see remove()
   */
  public static function update($files) {
    foreach($files as $file) {
      if ($file instanceof CrackupDirectory) {
        self::update($file);
      }
      elseif ($file instanceof CrackupFile) {
        self::debug($file->getName());
        
        $remoteFile = self::$remote.'/crackup_'.$file->getNameHash();
        $tempFile   = self::getTempFile();        
        
        // Create a compressed (and encrypted, if necessary) temporary file.
        if (defined('CRACKUP_NOGPG')) {
          self::compressFile($file->getName(), $tempFile);
        }
        else {
          self::encryptFile($file->getName(), $tempFile);
        }
          
        // Upload the temporary file to the remote location, then delete it.
        try {
          self::$driver->put($tempFile, $remoteFile);
        }
        catch (Exception $e) {
          @unlink($tempFile);
          self::error('Unable to upload file: '.$file->getName());
        }
        
        @unlink($tempFile);
      }
    }
  }
  
  /**
   * Brings the remote file index up to date with the local one.
   */
  public static function updateRemoteIndex() {
    $indexFile  = self::getTempFile();
    $remoteFile = self::$remote.'/.crackup_index';
    
    if (!file_put_contents($indexFile, serialize(self::$localFiles))) {
      self::error('Unable to write temporary file');
    }
    
    $oldFile   = $indexFile;
    $indexFile = self::getTempFile();
      
    if (defined('CRACKUP_NOGPG')) {
      self::compressFile($oldFile, $indexFile);
    }
    else {
      self::encryptFile($oldFile, $indexFile);
    }
    
    @unlink($oldFile);

    $success = false;
    
    while(!$success) {
      try {
        $success = self::$driver->put($indexFile, $remoteFile);
      }
      catch (Exception $e) {
        $tryAgain = self::prompt('Unable to update remote index. Try again? '.
            '(y/n)');
        
        if (strtolower($tryAgain) != 'y') {
          @unlink($indexFile);
          exit(-1);
        }
      }
    }
    
    @unlink($indexFile);
  }
}
?>

File: lib/crackup.rb
ENV['PATH'] = "#{File.dirname(__FILE__)};#{ENV['PATH']}"

require 'crackup/errors'
require 'crackup/dirobject'
require 'crackup/driver'
require 'crackup/fileobject'
require 'tempfile'
require 'zlib'

module Crackup

  GPG_DECRYPT = 'echo :passphrase | gpg --batch --quiet --no-tty --no-secmem-warning --cipher-algo aes256 --compress-algo bzip2 --passphrase-fd 0 --output :output_file :input_file'
  GPG_ENCRYPT = 'echo :passphrase | gpg --batch --quiet --no-tty --no-secmem-warning --cipher-algo aes256 --compress-algo bzip2 --passphrase-fd 0 --output :output_file --symmetric :input_file'
  
  attr_accessor :driver, :local_files, :options, :remote_files
  
  # Reads _infile_ and compresses it to _outfile_ using zlib compression.
  def self.compress_file(infile, outfile)
    File.open(infile, 'rb') do |input|
      Zlib::GzipWriter.open(outfile, 9) do |output|
        while data = input.read(1048576) do
          output.write(data)
        end
      end
    end
    
  rescue => e
    raise Crackup::CompressionError, "Unable to compress #{infile}: #{e}"
  end
  
  # Prints _message_ to +stdout+ if verbose mode is enabled.
  def self.debug(message)
    puts message if @options[:verbose] || $VERBOSE
  end
  
  # Reads _infile_ and decompresses it to _outfile_ using zlib compression.
  def self.decompress_file(infile, outfile)
    Zlib::GzipReader.open(infile) do |input|
      File.open(outfile, 'wb') do |output|
        while data = input.read(1048576) do
          output.write(data)
        end
      end
    end
  
  rescue => e
    raise Crackup::CompressionError, "Unable to decompress #{infile}: #{e}"
  end
  
  # Calls GPG to decrypt _infile_ to _outfile_.
  def self.decrypt_file(infile, outfile)
    File.delete(outfile) if File.exist?(outfile)

    gpg_command = String.new(GPG_DECRYPT)
    gpg_command.gsub!(':input_file', escapeshellarg(infile))
    gpg_command.gsub!(':output_file', escapeshellarg(outfile))
    gpg_command.gsub!(':passphrase', escapeshellarg(@options[:passphrase]))
    
    unless system(gpg_command)
      raise Crackup::EncryptionError, "Unable to decrypt file: #{infile}"
    end
  end
  
  def self.driver
    return @driver
  end
  
  # Calls GPG to encrypt _infile_ to _outfile_.
  def self.encrypt_file(infile, outfile)
    File.delete(outfile) if File.exist?(outfile)

    gpg_command = String.new(GPG_ENCRYPT)
    gpg_command.gsub!(':input_file', escapeshellarg(infile))
    gpg_command.gsub!(':output_file', escapeshellarg(outfile))
    gpg_command.gsub!(':passphrase', escapeshellarg(@options[:passphrase]))
    
    unless system(gpg_command)
      raise Crackup::EncryptionError, "Unable to encrypt file: #{infile}"
    end
  end
  
  # Prints the specified _message_ to +stderr+ and exits with an error
  # code of 1.
  def self.error(message)
    abort "#{APP_NAME}: #{message}"
  end
  
  # Wraps _arg_ in single quotes, escaping any single quotes contained therein,
  # thus making it safe for use as a shell argument.
  def self.escapeshellarg(arg)
    return "'#{arg.gsub("'", "\\'")}'"
  end
  
  # Gets an array of files in the remote file index whose local paths match
  # _pattern_.
  def self.find_remote_files(pattern)
    files = []
    pattern.chomp!('/')
    
    @remote_files.each do |name, file|
      if File.fnmatch?(pattern, file.name)
        files << file
        next
      end

      next unless file.is_a?(Crackup::DirectoryObject)
      files += file.find(pattern)
    end
    
    return files
  end
  
  # Gets a flat array of filenames from _files_, which may be either a Hash
  # or a Crackup::FileSystemObject.
  def self.get_list(files)
    list = []
  
    if files.is_a?(Hash)
      files.each_value {|value| list += get_list(value) }
    elsif files.is_a?(Crackup::DirectoryObject)
      list += get_list(files.children)
    elsif files.is_a?(Crackup::FileObject)
      list << files.name 
    end
    
    return list
  end

  # Gets a Hash of Crackup::FileSystemObjects representing the files and
  # directories on the local system in the locations specified by the array of
  # filenames in <tt>options[:from]</tt>.
  def self.get_local_files
    local_files = {}
    
    @options[:from].each do |filename|
      next unless File.exist?(filename = filename.chomp('/'))
      next if local_files.has_key?(filename)
      
      # Skip this file if it's in the exclusion list.
      unless @options[:exclude].nil?
        next if @options[:exclude].any? do |pattern|
          File.fnmatch?(pattern, filename)
        end
      end
      
      if File.directory?(filename)
        debug "--> #{filename}"
        local_files[filename] = Crackup::DirectoryObject.new(filename)
      elsif File.file?(filename)
        debug "--> #{filename}"
        local_files[filename] = Crackup::FileObject.new(filename)
      end
    end
    
    return local_files
  end
  
  # Gets a Hash of Crackup::FileSystemObjects present at the remote location.
  def self.get_remote_files(url)
    tempfile = get_tempfile()
    
    # Download the index file.
    begin
      @driver.get(url + '/.crackup_index', tempfile)
    rescue => e
      return {}
    end
    
    # Decompress/decrypt the index file.
    oldfile  = tempfile
    tempfile = get_tempfile()
    
    if @options[:passphrase].nil?
      begin
        decompress_file(oldfile, tempfile)
      rescue => e
        raise Crackup::IndexError, "Unable to decompress index file. Maybe " +
            "it's encrypted?"
      end
    else
      begin
        decrypt_file(oldfile, tempfile)
      rescue => e
        raise Crackup::IndexError, "Unable to decrypt index file."
      end
    end
    
    # Load the index file.
    file_list = {}

    begin
      File.open(tempfile, 'rb') {|file| file_list = Marshal.load(file) }
    rescue => e
      raise Crackup::IndexError, "Remote index is invalid!"
    end
    
    unless file_list.is_a?(Hash)
      raise Crackup::IndexError, "Remote index is invalid!"
    end
    
    return file_list
  end
  
  # Gets an Array of Crackup::FileSystemObjects representing files and
  # directories that exist at the remote location but no longer exist at the
  # local location.
  def self.get_removed_files(local_files, remote_files)
    removed = []

    remote_files.each do |name, remotefile|
      unless local_files.has_key?(name)
        removed << remotefile
        next
      end

      localfile = local_files[name]
      
      if remotefile.is_a?(Crackup::DirectoryObject) && 
          localfile.is_a?(Crackup::DirectoryObject)
        removed += get_removed_files(localfile.children, remotefile.children)
      end
    end
    
    return removed
  end
  
  # Creates a new temporary file in the system's temporary directory and returns
  # its name. All temporary files will be deleted when the program exits.
  def self.get_tempfile
    tempfile = Tempfile.new('.crackup')
    tempfile.close
    
    return tempfile.path
  end

  # Gets an Array of Crackup::FileSystemObjects representing files and
  # directories that are new or have been modified at the local location and
  # need to be updated at the remote location.
  def self.get_updated_files(local_files, remote_files)
    updated = []
    
    local_files.each do |name, localfile|
      # Add the file to the list if it doesn't exist at the remote location.
      unless remote_files.has_key?(name)
        updated << localfile
        next
      end
      
      remotefile = remote_files[name]
      
      if localfile.is_a?(Crackup::DirectoryObject) && 
          remotefile.is_a?(Crackup::DirectoryObject)
        # Add to the list all updated files contained in the directory and its 
        # subdirectories.
        updated += get_updated_files(localfile.children, remotefile.children)
      elsif localfile.is_a?(Crackup::FileObject) && 
          remotefile.is_a?(Crackup::FileObject)
        # Add the file to the list if the local file has been modified.
        unless localfile.file_hash == remotefile.file_hash
          updated << localfile
        end
      end
    end
    
    return updated
  end
  
  def self.options
    return @options
  end

  # Prints _message_ to +stdout+ and waits for user input, which is then
  # returned.
  def self.prompt(message)
    puts message + ': '
    return $stdin.gets
  end
  
  # Deletes each Crackup::FileSystemObject specified in the _files_ array from
  # the remote location.
  def self.remove_files(files)
    files.each do |file|
      file.remove
    end
  end
  
  # Uploads each Crackup::FileSystemObject specified in the _files_ array to the
  # remote location.
  def self.update_files(files)
    files.each do |file|
      file.update
    end
  end
  
  # Brings the remote file index up to date with the local one.
  def self.update_remote_index
    tempfile   = get_tempfile()
    remotefile = @options[:to] + '/.crackup_index'
  
    File.open(tempfile, 'wb') {|file| Marshal.dump(@local_files, file) }
    
    oldfile  = tempfile
    tempfile = get_tempfile()
    
    if @options[:passphrase].nil?
      compress_file(oldfile, tempfile)
    else
      encrypt_file(oldfile, tempfile)
    end
    
    begin
      success = @driver.put(remotefile, tempfile)
    rescue => e
      tryagain = prompt('Unable to update remote index. Try again? (y/n)')
    
      retry if tryagain.downcase == 'y'
      raise Crackup::IndexError, "Unable to update remote index: #{e}"
    end
  end
  
end
